erpnext/stock/doctype/serial_no/serial_no.py

```diff
@@ -64,6 +64,37 @@ def __init__(self, *args, **kwargs):
 		super(SerialNo, self).__init__(*args, **kwargs)
 		self.via_stock_ledger = False
 
+	def autoname(self):
+		self.name = self.serial_no
+
+		if frappe.db.get_single_value("Stock Settings", "allow_duplicate_serial_nos"):
+			self.validate_duplicate_in_same_item()
+			self.name = self.autoname_duplicate_serial_no()
```

Comment:
```suggestion
		if frappe.db.get_single_value("Stock Settings", "allow_duplicate_serial_nos"):
			self.validate_duplicate_in_same_item()
			self.name = self.autoname_duplicate_serial_no()
		else:
			self.name = self.serial_no
```

A bit more readable, IMO. (I don't have to worry about whether the default name could influence the duplicate name.)